system/security/shared/caching.cpp

```diff
@@ -672,20 +680,42 @@ bool CPermissionsCache::queryPermsManagedFileScope(ISecUser& sec_user, const cha
 
 SecAccessFlags CPermissionsCache::queryDefaultPermission(ISecUser& user)
 {
-    const char *username = user.getName();
 
-    SecAccessFlags defaultPermission = SecAccess_None;
-    auto it = m_userDefaultFileScopePermissions.find(username);
-    if (it == m_userDefaultFileScopePermissions.end())
+    if (m_useLegacyDefaultFileScopePermissionCache)
     {
-        defaultPermission = m_secMgr->queryDefaultPermission(user);
-        std::string userName(username);
-        m_userDefaultFileScopePermissions.insert(std::pair<std::string, SecAccessFlags>(userName, defaultPermission));
-        DBGLOG("Added user '%s' to default file scope permissions with access %s(%d)", username, getSecAccessFlagName(defaultPermission), defaultPermission);
+        if (m_defaultPermission == SecAccess_Unknown)
+        {
+            if (m_secMgr)
+                m_defaultPermission = m_secMgr->queryDefaultPermission(user);
+            else
+                m_defaultPermission = SecAccess_None;
+
+            DBGLOG("Legacy default file scope permission set to %s(%d) for all users, based on User '%s'", getSecAccessFlagName(m_defaultPermission),
+                   m_defaultPermission, user.getName());
+        }
+        return m_defaultPermission;
     }
-    else
+
+    const char *username = user.getName();
+    SecAccessFlags defaultPermission = SecAccess_None;
+    if (m_secMgr)
     {
-        defaultPermission = it->second;
+        ReadLockBlock defaultScopesReadLock(m_defaultScopesRWLock);
+        auto it = m_userDefaultFileScopePermissions.find(username);
+        if (it == m_userDefaultFileScopePermissions.end())
+        {
+            defaultScopesReadLock.clear();
+            WriteLockBlock defaultScopesWriteLock(m_defaultScopesRWLock);
+            defaultPermission = m_secMgr->queryDefaultPermission(user);
+            std::string userName(username);
```

Comment:
at least in my setup, this line is not necessary, i.e. you can pass const char * to the emplace and it will implicitly construct a std::string

Reply:
The _find_, in my experience, has to construct an object of the key type.  So, I left the username std::string variable, but cleaned it up a bit.